In [1]:
import pandas as pd

In [86]:
import re

In [68]:
patients=pd.read_csv('patients.csv')
treatment=pd.read_csv('treatments.csv')
treatment2=pd.read_csv('treatments_cut.csv')
adverse=pd.read_csv('adverse_reactions.csv')

In [4]:
print(patients.shape)
print(treatment.shape)
print(adverse.shape)

(503, 14)
(280, 7)
(34, 3)


In [61]:
treatment

given_name       surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika      jindrová  41u - 48u          -         7.63       7.20   
1       elliot    richardson          -  40u - 45u         7.56       7.09   
2     yukitaka      takenaka          -  39u - 36u         7.68       7.25   
3         skye    gormanston  33u - 36u          -         7.97       7.62   
4       alissa        montez          -  33u - 29u         7.78       7.46   
..         ...           ...        ...        ...          ...        ...   
275     albina      zetticci  45u - 51u          -         7.93       7.73   
276       john   teichelmann          -  49u - 49u         7.90       7.58   
277     mathea       lillebø  23u - 36u          -         9.04       8.67   
278     vallie        prince  31u - 38u          -         7.64       7.28   
279     samúel  guðbrandsson  53u - 56u          -         8.00       7.64   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
275          0.20  
276           NaN  
277          0.37  
278          0.36  
279          0.36  

[280 rows x 7 columns]

In [14]:
adverse.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [9]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
patient_id      503 non-null int64
assigned_sex    503 non-null object
given_name      503 non-null object
surname         503 non-null object
address         491 non-null object
city            491 non-null object
state           491 non-null object
zip_code        491 non-null float64
country         491 non-null object
contact         491 non-null object
birthdate       503 non-null object
weight          503 non-null float64
height          503 non-null int64
bmi             503 non-null float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


In [12]:
treatment.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    171 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


In [13]:
adverse.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 34 entries, 0 to 33
Data columns (total 3 columns):
given_name          34 non-null object
surname             34 non-null object
adverse_reaction    34 non-null object
dtypes: object(3)
memory usage: 896.0+ bytes


In [10]:
patients[patients['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

In [14]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000

In [15]:
treatment.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

In [19]:
patients.sample(5)

patient_id assigned_sex given_name     surname             address  \
296         297       female        Chỉ         Lâm                 NaN   
13           14       female   Anenechi       Chidi    826 Broad Street   
177         178       female     Monika      Lončar   377 Norman Street   
5             6         male     Rafael       Costa  1140 Willis Avenue   
457         458       female      Mevla  Grabljevec   2645 Moore Avenue   

              city    state  zip_code        country  \
296            NaN      NaN       NaN            NaN   
13      Birmingham       AL   35203.0  United States   
177    Los Angeles       CA   90014.0  United States   
5    Daytona Beach  Florida   32114.0  United States   
457     Fort Worth       TX   76102.0  United States   

                                        contact  birthdate  weight  height  \
296                                         NaN  5/14/1990   181.1      63   
13   AnenechiChidi@armyspy.com+1 (205) 417-8095   3/7/1961   228.4      67   
177          323-327-4721MonikaLoncar@rhyta.com  7/19/1927   203.7      63   
5      386-334-5237RafaelCardosoCosta@gustr.com  8/31/1931   183.9      70   
457     817-777-5289MevlaGrabljevec@teleworm.us  12/2/1984   215.8      66   

      bmi  
296  32.1  
13   35.8  
177  36.1  
5    26.4  
457  34.8

In [22]:
patients[patients['address'].duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

### Quality

[Completeness, Validity, Accuracy, Consistency]

#### `Patients`

- Zip code col is stored in decimal format[V]
- Certain zip codes of 4 digits[V] 
- For the state col sometimes full names are use and other times short forms are used[IC]
- Missing values for contact details of 12 patients[C]
- Incorrect data types for assigned_sex and birthdate cols[?]
- Erroneous data for weights and heights[Accuracy]
- Inconsistent way of storing phone numbers in contact col[IC]
- Duplicate and not useful data in dataset[IC]

#### `Treatments`

- hba1c_change col has missing values[C]
- Novodra and Auralin cols has u attached to numbers[V]
- Only 280 rows are there, instead of 350[C]
- Name and surname col is in lower case[IC]
- Incorrect data type for auralin and novadra cols[V]
- hba1c_change 9 instead of 4[Accuracy]
- use of dashes instead of Nan in aurolin and novodra cols[IC]

#### `Adverse Reactions`

- Name and surname col is in lower case[IC]

### Tidiness

#### `Patients`

- Phone and email together in one col i.e. contact

#### `Treatments`

- 3 variables stored in auralin and novodra cols

#### `Adverse Reactions`

- No point of using this table


# Cleaning Data

You should follow the order provided below:

- Completeness Issues
- Tidiness Issues
- Remaining Data quality issues like validity, accuracy and consistency

### Steps involved in Data cleaning

- Define
- Code 
- Test

`Always make sure to create a copy of your pandas dataframe before you start the cleaning process`

In [95]:
# Creating copies

patients_copy=patients.copy()
treatment_copy=treatment.copy()
adverse_copy=adverse.copy()
treatment2_copy=treatment2.copy()

In [9]:
patients_copy[patients_copy['address'].isnull()]

patient_id assigned_sex given_name      surname address city state  \
209         210       female     Lalita  Eldarkhanov     NaN  NaN   NaN   
219         220         male         Mỹ        Quynh     NaN  NaN   NaN   
230         231       female  Elisabeth      Knudsen     NaN  NaN   NaN   
234         235       female    Martina    Tománková     NaN  NaN   NaN   
242         243         male       John      O'Brian     NaN  NaN   NaN   
249         250         male   Benjamin       Mehler     NaN  NaN   NaN   
257         258         male        Jin         Kung     NaN  NaN   NaN   
264         265       female   Wafiyyah       Asfour     NaN  NaN   NaN   
269         270       female     Flavia   Fiorentino     NaN  NaN   NaN   
278         279       female   Generosa        Cabán     NaN  NaN   NaN   
286         287         male      Lewis         Webb     NaN  NaN   NaN   
296         297       female        Chỉ          Lâm     NaN  NaN   NaN   

     zip_code country contact   birthdate  weight  height   bmi  
209       NaN     NaN     NaN   8/14/1950   143.4      62  26.2  
219       NaN     NaN     NaN    4/9/1978   237.8      69  35.1  
230       NaN     NaN     NaN   9/23/1976   165.9      63  29.4  
234       NaN     NaN     NaN    4/7/1936   199.5      65  33.2  
242       NaN     NaN     NaN   2/25/1957   205.3      74  26.4  
249       NaN     NaN     NaN  10/30/1951   146.5      69  21.6  
257       NaN     NaN     NaN   5/17/1995   231.7      69  34.2  
264       NaN     NaN     NaN   11/3/1989   158.6      63  28.1  
269       NaN     NaN     NaN   10/9/1937   175.2      61  33.1  
278       NaN     NaN     NaN  12/16/1962   124.3      69  18.4  
286       NaN     NaN     NaN    4/1/1979   155.3      68  23.6  
296       NaN     NaN     NaN   5/14/1990   181.1      63  32.1

## a. Completeness Issues

- Missing values for contact details of 12 patients
- Only 280 rows are there, instead of 350
- hba1c_change col has missing values

# Define

In [ ]:
#Merging the rows of treatments_cut.csv with treatments to get 350 rows

# Code

In [76]:
treatments_copy=pd.concat([treatment_copy,treatment2_copy],ignore_index=True,sort=False)

# Test

In [77]:
treatments_copy

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika    jindrová  41u - 48u          -         7.63       7.20   
1       elliot  richardson          -  40u - 45u         7.56       7.09   
2     yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3         skye  gormanston  33u - 36u          -         7.97       7.62   
4       alissa      montez          -  33u - 29u         7.78       7.46   
..         ...         ...        ...        ...          ...        ...   
345     rovzan     kishiev  32u - 37u          -         7.75       7.41   
346      jakob    jakobsen          -  28u - 26u         7.96       7.51   
347      bernd   schneider  48u - 56u          -         7.74       7.44   
348      berta  napolitani          -  42u - 44u         7.68       7.21   
349     armina       sauvé  36u - 46u          -         7.86       7.40   

     hba1c_change  
0             NaN  
1            0.97  
2             NaN  
3            0.35  
4            0.32  
..            ...  
345          0.34  
346          0.95  
347          0.30  
348           NaN  
349           NaN  

[350 rows x 7 columns]

#### Define

In [ ]:
# Subtract hba1c_start from hba1c_end to get hba1c_change

#### Code

In [6]:
treatment_copy['hba1c_change']=treatment_copy['hba1c_start']-treatment_copy['hba1c_end']

#### Test

In [7]:
treatment_copy.info()

treatment_copy.describe()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
given_name      280 non-null object
surname         280 non-null object
auralin         280 non-null object
novodra         280 non-null object
hba1c_start     280 non-null float64
hba1c_end       280 non-null float64
hba1c_change    280 non-null float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    280.000000
mean      7.985929    7.589286      0.396643
std       0.568638    0.569672      0.060168
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.350000
50%       7.800000    7.420000      0.390000
75%       7.970000    7.570000      0.440000
max       9.950000    9.580000      0.540000

## b. Tidiness Issues

- Phone + email in contact col `patients`
- Novodra and Aurolin col `treatment`
- Merge the `adverse` table to `treatment` table

#### Define

In [36]:
# Novodra and Aurolin using the melt function

#### Code

In [25]:
treatment_copy=pd.melt(treatment_copy, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'],
                           var_name='treatment', value_name='dose')

In [26]:
treatment_copy=treatment_copy[treatment_copy['dose']!='-']

In [27]:
treatment_copy['start'],treatment_copy['end']=treatment_copy['dose'].str.split('-').str

In [28]:
treatment_copy.drop(columns={'dose'}, inplace=True)

In [30]:
treatment_copy

given_name      surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0     veronika     jindrová         7.63       7.20           NaN   auralin   
3         skye   gormanston         7.97       7.62          0.35   auralin   
6       sophia       haugen         7.65       7.27          0.38   auralin   
7        eddie       archer         7.89       7.55          0.34   auralin   
9         asia      woźniak         7.76       7.37           NaN   auralin   
..         ...          ...          ...        ...           ...       ...   
548     yasmin       araujo         7.82       7.36          0.96   novodra   
549     hiromu     horikawa         7.77       7.28           NaN   novodra   
551        leo       vieira         7.74       7.36           NaN   novodra   
552     steven          roy         7.87       7.43          0.94   novodra   
556       john  teichelmann         7.90       7.58           NaN   novodra   

    start   end  
0    41u    48u  
3    33u    36u  
6    37u    42u  
7    31u    38u  
9    30u    36u  
..    ...   ...  
548  51u    54u  
549  47u    46u  
551  30u    33u  
552  41u    43u  
556  49u    49u  

[280 rows x 8 columns]

#### Test

In [48]:
treatment_copy.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  start   end  
0  41u    48u  
3  33u    36u  
6  37u    42u  
7  31u    38u  
9  30u    36u

# Define

In [53]:
#Merging the adverse table to treatment table

# Code

In [80]:
treatmentt=treatments_copy.merge(adverse_copy,how='outer')

# Test

In [81]:
treatmentt

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0     veronika    jindrová  41u - 48u          -         7.63       7.20   
1       elliot  richardson          -  40u - 45u         7.56       7.09   
2     yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3         skye  gormanston  33u - 36u          -         7.97       7.62   
4       alissa      montez          -  33u - 29u         7.78       7.46   
..         ...         ...        ...        ...          ...        ...   
345     rovzan     kishiev  32u - 37u          -         7.75       7.41   
346      jakob    jakobsen          -  28u - 26u         7.96       7.51   
347      bernd   schneider  48u - 56u          -         7.74       7.44   
348      berta  napolitani          -  42u - 44u         7.68       7.21   
349     armina       sauvé  36u - 46u          -         7.86       7.40   

     hba1c_change           adverse_reaction  
0             NaN                        NaN  
1            0.97               hypoglycemia  
2             NaN                        NaN  
3            0.35                        NaN  
4            0.32                        NaN  
..            ...                        ...  
345          0.34                        NaN  
346          0.95               hypoglycemia  
347          0.30                        NaN  
348           NaN  injection site discomfort  
349           NaN                        NaN  

[350 rows x 8 columns]

# Define

In [ ]:
#Phone + email in contact col patients

# Code

In [96]:
impossible=patients_copy[patients_copy['address'].isnull()]
patients_copy.dropna(inplace=True)
e=[]
n=[]
for i in patients_copy['contact']:
    p=re.compile(r'[a-zA-Z][a-zA-Z0-9.-]+[@][a-zA-Z0-9-]+\.[a-zA-Z0-9.-]+[a-zA-Z]')
    n.append(i.replace(p.findall(i)[0],''))
    e.append(p.findall(i)[0])
patients_copy['Email']=e
patients_copy['Ph']=n
patients_copy=pd.concat([patients_copy,impossible]).reset_index()

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:11: FutureWarning: Sorting because non-concatenation axis is not aligned. A future version
of pandas will change to not sort by default.

To accept the future behavior, pass 'sort=False'.

To retain the current behavior and silence the warning, pass 'sort=True'.

  # This is added back by InteractiveShellApp.init_path()


In [97]:
patients_copy=patients_copy.drop(columns=['index','contact'])

# test

In [98]:
patients_copy.head(15)

Email                 Ph  \
0          ZoeWellish@superrito.com       951-719-9170   
1              PamelaSHill@cuvox.de  +1 (217) 569-3204   
2              JaeMDebord@gustr.com       402-363-6804   
3         PhanBaLiem@jourrapide.com  +1 (732) 636-8246   
4               TimNeudorf@cuvox.de       334-515-7487   
5      RafaelCardosoCosta@gustr.com       386-334-5237   
6             MaryBAdams@einrot.com       775-533-5933   
7            XiuxiuChang@einrot.com     1 408 778 3236   
8       DavidGustafsson@armyspy.com       816-265-9578   
9   SophieCabreraIbarra@teleworm.us     1 718 795 9124   
10       SandyGunnarsson@dayrep.com       906-478-8949   
11     Abdul-NurMummarIsa@rhyta.com     1 931 207 0839   
12    OmeokachieIbeamaka@einrot.com       434-509-2614   
13        AnenechiChidi@armyspy.com  +1 (205) 417-8095   
14            AsiaWozniak@rhyta.com       918-712-3469   

                       address assigned_sex   birthdate   bmi  \
0         576 Brown Bear Drive       female   7/10/1976  19.6   
1    2370 University Hill Road       female    4/3/1967  19.2   
2        1493 Poling Farm Road         male   2/19/1980  24.8   
3          2335 Webster Street         male   7/26/1951  31.7   
4         1428 Turkey Pen Lane         male   2/18/1928  26.1   
5           1140 Willis Avenue         male   8/31/1931  26.4   
6             3145 Sheila Lane       female  11/19/1969  24.3   
7   2687 Black Oak Hollow Road       female   8/13/1958  30.9   
8           1790 Nutter Street         male    3/6/1937  26.5   
9            3303 Anmoore Road       female   12/3/1930  33.4   
10          87 Wood Duck Drive       female   7/16/1974  36.4   
11      1092 Farm Meadow Drive         male    2/3/1954  31.5   
12          2544 Worley Avenue         male    8/5/1957  33.1   
13            826 Broad Street       female    3/7/1961  35.8   
14       4970 Heather Sees Way       female   8/15/1997  18.6   

                 city        country  given_name  height  patient_id  \
0   Rancho California  United States         Zoe      66           1   
1           Armstrong  United States      Pamela      66           2   
2                York  United States         Jae      71           3   
3          Woodbridge  United States        Liêm      70           4   
4              Dothan  United States         Tim      27           5   
5       Daytona Beach  United States      Rafael      70           6   
6             Burbank  United States        Mary      65           7   
7         Morgan Hill  United States      Xiuxiu      60           8   
8         Kansas City  United States       Dsvid      66           9   
9            New York  United States      Sophie      64          10   
10            Rudyard  United States       Sandy      62          11   
11          Brentwood  United States   Abdul-Nur      73          12   
12          Lynchburg  United States  Omeokachie      69          13   
13         Birmingham  United States    Anenechi      67          14   
14              Tulsa  United States        Asia      65          15   

         state     surname  weight  zip_code  
0   California     Wellish   121.7   92390.0  
1     Illinois        Hill   118.8   61812.0  
2     Nebraska      Debord   177.8   68467.0  
3           NJ        Phan   220.9    7095.0  
4           AL     Neudorf   192.3   36303.0  
5      Florida       Costa   183.9   32114.0  
6           NV       Adams   146.3   84728.0  
7           CA       Chang   158.0   95037.0  
8           MO  Gustafsson   163.9   64105.0  
9     New York     Cabrera   194.7   10011.0  
10          MI  Gunnarsson   199.3   49780.0  
11          TN         Isa   238.7   37027.0  
12          VA    Ibeamaka   224.2   24504.0  
13          AL       Chidi   228.4   35203.0  
14          OK     Woźniak   112.0   74105.0

### What is Assessing Data?

The data analysis process consists of the following steps

1. Asking Questions
2. Data Wrangling<br><br>
    a. Gathering Data
            i.   CSV files
            ii.  APIs
            iii. Web Scraping
            iv.  Databases
    b. <b>Assessing Data</b><br>
    c. Cleaning Data<br><br>
3. Exploratory Data Analysis
4. Drawing Conclusion
5. Comunicating Results
    

### Assessing Data

In this step, the data is to be understood more deeply. Before implementing methods to clean it, you will definitely need to have a better idea about what the data is about.

### Types of Unclean Data

There are 2 kinds of unclean data

1. <u>Dirty Data</u> (Data with Quality issues):<br>
    Dirty data, also known as low quality data. Low quality data has content issues.<br>
2. <u>Messy Data</u> (Data with tidiness issues)<br>
    Messy data, also known as untidy data. Untidy data has structural issues.

### 1. Dirty Data

When the data has following kinds of issues:

1. Duplicated data
2. Missing Data
3. Corrupt Data
4. Inaccurate Data

This kind of data is known as Dirty data

### 2. Messy Data

When the data has structural issue is it known as Messy or untidy data.

Tidy data has the following properties:

1. Each variable forms a column
2. Each observation forms a row
3. Each observational unit forms a table

If a data violates any one of the above it is known as untidy data

Example:

<img src="untidy.png">

### Types of Assessment

There are 2 types of assessment styles

1. Manual - Looking through the data manually in google sheets
2. Programmatic - By using pandas functions such as info(), describe() or sample()

### Steps in Assessment

There are 2 steps involved in Assessment

1. Discover
2. Document

### Data Quality Dimensions

1. Completeness Issues
2. Validity Issues
3. Accuracy Issues
4. Consistency Issues